common/src/main/java/org/vivecraft/api/client/VivecraftClientAPI.java

```diff
@@ -0,0 +1,217 @@
+     */
+    @Beta
+    boolean isVanillaRenderPass();
+
+    /**
+     * @return The currently active world scale.
+     */
+    float getWorldScale();
+
+    /**
+     * Returns the history of VR poses for the player for the HMD. Will return null if the player isn't
+     * in VR.
+     *
+     * @return The historical VR data for the player's HMD, or null if the player isn't in VR.
+     */
+    @Nullable
+    VRPoseHistory getHistoricalVRHMDPoses();
+
+    /**
+     * Returns the history of VR poses for the player for a controller. Will return null if the player isn't
+     * in VR.
+     *
+     * @param controller The controller number to get, with 0 being the primary controller.
+     * @return The historical VR data for the player's controller, or null if the player isn't in VR.
+     */
+    @Nullable
+    VRPoseHistory getHistoricalVRControllerPoses(int controller);
+
+    /**
+     * Returns the history of VR poses for the player for a controller. Will return null if the player isn't
+     * in VR.
+     *
+     * @param hand The hand to get controller history for.
+     * @return The historical VR data for the player's controller, or null if the player isn't in VR.
+     */
+    @Nullable
+    default VRPoseHistory getHistoricalVRControllerPoses(InteractionHand hand) {
+        return getHistoricalVRControllerPoses(hand.ordinal());
+    }
+
+    /**
+     * Returns the history of VR poses for the player for the primary controller. Will return null if the
+     * player isn't in VR.
+     *
+     * @return The historical VR data for the player's primary controller, or null if the player isn't in VR.
+     */
+    @Nullable
+    default VRPoseHistory getHistoricalVRController0Poses() {
+        return getHistoricalVRControllerPoses(0);
+    }
+
+    /**
+     * Returns the history of VR poses for the player for the secondary controller. Will return null if the
+     * player isn't in VR.
+     *
+     * @return The historical VR data for the player's secondary controller, or null if the player isn't in VR.
+     */
+    @Nullable
+    default VRPoseHistory getHistoricalVRController1Poses() {
+        return getHistoricalVRControllerPoses(1);
+    }
+
+    /**
+     * Opens or closes Vivecraft's keyboard. Will fail silently if the user isn't in VR or if the keyboard's new state
+     * is the same as the old.
+     *
+     * @param isNowOpen Whether the keyboard should now be open. If false, the keyboard will attempt to close.
+     * @return Whether the keyboard is currently showing after attempting to open/close it.
+     */
+    boolean setKeyboardState(boolean isNowOpen);
+
+    /**
+     * Opens Vivecraft's keyboard to allow VR users to easily input text. Will fail silently if the user isn't in VR or
+     * if the keyboard is already open.
+     *
+     * @return Whether the keyboard is currently showing after attempting to open it.
+     */
+    default boolean openKeyboard() {
```

Comment:
Are these 2 really needed? just setKeyboardState(boolean isNowOpen) should be enough imo.

Reply:
In retrospect, completely agree with you there.

Reply:
`openKeyboard` and `closeKeyboard` removed in https://github.com/hammy275/VivecraftMod/commit/0f667b0f452be1b68a4e418d1b726d490121c1a7